### PR TITLE
adding absolute timestamp to ASC reader

### PR DESCRIPTION
**Pull request by cperkulator**
needed timestamps to be absolute so I added this. 

Probably needs some error handling.  Will get to it when I have the chance.

### CODE REVIEW

File: can/io/asc.py
Comment (christiansandberg) [inline can/io/asc.py]: What happens if the parsing fails? Should we catch some exception?

File: can/io/asc.py
Comment (bessman) [inline can/io/asc.py]:
```suggestion
```

File: can/io/asc.py
Comment (bessman) [inline can/io/asc.py]:
```suggestion
                try:
                    _, _, start_time = temp.split(None, 2)
                    start_time = datetime.strptime(start_time, "%a %b %m %I:%M:%S.%f %p %Y").timestamp()
                except ValueError:
                    start_time = None
```

File: can/io/asc.py
Comment (bessman) [inline can/io/asc.py]:
```suggestion

```

File: can/io/asc.py
Comment (bessman) [inline can/io/asc.py]:
```suggestion
            timestamp = float(timestamp) + (start_time if start_time is not None else 0)
```

File: can/io/asc.py
Comment (bessman) [reply can/io/asc.py]: Both of these lines could raise ValueError, if `temp` contains fewer than two whitespaces, or if it does not match the specified date format.

Comment (felixdivo) [reply can/io/asc.py]: One idea would be to raise an error here, and maybe let the user suppress it via a param to the constructor.

**Comment by felixdivo** [comment]
This might need a rebase due to #741 having been merged two days ago.

File: can/io/asc.py
Comment (christiansandberg) [reply can/io/asc.py]: I think we could fall back to relative timestamps since absolute timestamps is not guaranteed in either log files or buses.

Comment (felixdivo) [reply can/io/asc.py]: > I think we could fall back to relative timestamps since absolute timestamps is not guaranteed in either log files or buses.

Yeah, but maybe print a warning along the way.

**Comment by sou1hacker** [comment]
Was wondering if we can make this optional by a flag passed to LogReader? In most of our use cases, we would actually prefer the current timestamp format which is easier to read and co-relate back with the ascii logs file. Maybe even make the existing implementation the default approach?

**Comment by felixdivo** [comment]
@cperkulator Do you plan on having another look at this?

**Comment by cperkulator** [comment]
> @cperkulator Do you plan on having another look at this?

oops I really apologize. I'll get this rebased and try and wrap it up today/

File: can/io/asc.py
Comment (felixdivo) [inline can/io/asc.py:50]:
```suggestion
                     at 0.0) or `absolute` (starting at the system time).
```

File: can/io/asc.py
Comment (felixdivo) [inline can/io/asc.py:59]: I'd also name the argument `relative_timestamp` to keep it consistent with the attribute.

**Review by felixdivo — CHANGES_REQUESTED** [review]
Would you mind adding one test for each of the modes `relative` and `absolute`? That would make your PR really great!

File: test/data/test_CanMessage.asc
Comment (cperkulator) [inline test/data/test_CanMessage.asc:5]: looks like the tests files weren't made with the `ASCWriter` so I updated to make it right. Doesn't appear to mess with any tests.

File: test/logformats_test.py
Comment (cperkulator) [inline test/logformats_test.py:385]: @felixdivo Added just one test for absolute timestamp since relative is tested by the existing tests

File: test/data/test_CanMessage.asc
Comment (felixdivo) [reply test/data/test_CanMessage.asc:5]: Looks good. Looks like it was generated with a German locale.

File: test/logformats_test.py
Comment (felixdivo) [reply test/logformats_test.py:385]: Cool! That should suffice.

**Review by felixdivo — APPROVED** [review]

**Comment by felixdivo** [comment]
@Mergifyio update

**Comment by cperkulator** [comment]
Not sure what is up with codecov but according to tox, everything I added is covered.

**Comment by felixdivo** [comment]
> Not sure what is up with codecov but according to tox, everything I added is covered.

Yeah, just ignore it in this case. It's somewhat broken.

**Comment by felixdivo** [comment]
@cperkulator Somehow, there seem to be sudden errors on the develop branch. See [here](https://github.com/hardbyte/python-can/actions/runs/779056159). Do you know what's happening? They are probably related to this PR, right?

**Comment by cperkulator** [comment]
> @cperkulator Somehow, there seem to be sudden errors on the develop branch. See [here](https://github.com/hardbyte/python-can/actions/runs/779056159). Do you know what's happening? They are probably related to this PR, right?

They all passed before the merge right? I'll take a look.

**Comment by cperkulator** [comment]
> > @cperkulator Somehow, there seem to be sudden errors on the develop branch. See [here](https://github.com/hardbyte/python-can/actions/runs/779056159). Do you know what's happening? They are probably related to this PR, right?
> 
> They all passed before the merge right? I'll take a look.

@felixdivo alright figured it out.  Those failing tests create a header in the temporary ASC file:
image

Since this time is before 1970 (and thus not representable by POSIX time) then we get an `OSError` exception.

I'll submit a new PR to catch that error

**Comment by felixdivo** [comment]
Okay, sounds plausible. Thanks for figuring it out!

**Pull request merged**
